src/gc/collector.cpp

```diff
@@ -297,73 +565,24 @@ void markPhase() {
 #endif
     GC_TRACE_LOG("Starting collection %d\n", ncollections);
 
+    // Starting points of graph traversal.
     GC_TRACE_LOG("Looking at roots\n");
-    TraceStack stack(roots);
-    GCVisitor visitor(&stack);
+    getMarkPhaseRoots(visitor);
 
-    GC_TRACE_LOG("Looking at the stack\n");
-    threading::visitAllStacks(&visitor);
-
-    GC_TRACE_LOG("Looking at root handles\n");
-    for (auto h : *getRootHandles()) {
-        visitor.visit(h->value);
-    }
-
-    GC_TRACE_LOG("Looking at potential root ranges\n");
-    for (auto& e : potential_root_ranges) {
-        visitor.visitPotentialRange((void* const*)e.first, (void* const*)e.second);
+    if (also_free_classes) {
+        // If we allow freeing classes, we must make sure that even for unreachable objects,
```

Comment:
I still think there's a way to do this by ordering things during the sweep phase.  While sweeping we can check if the object we're about to free is a class and add it to a list, then when we're done freeing instances, we do a walk of that list freeing the classes.

If the hit of doing the additional linear walk of just the classes is still really slow, we can do it every few collections, but it would be better to base the decision to collect classes on collection frequency instead of number of classes allocated, since the former goes up with heap occupancy/pressure and the latter doesn't -- i.e. I allocate ALLOC_CLASSES_PER_CLASS_COLLECTION-1 classes, only keep a reference to 1 of them, then fill up the heap over and over.